### PR TITLE
Adds toSet to create sets from iterables

### PR DESCRIPTION
**Pull request by beef331**
Replicates the `toHashSet` proc for normal sets.

### CODE REVIEW

File: lib/system/setops.nim [hunk 1]
```diff
@@ -9,6 +9,17 @@ func incl*[T](x: var set[T], y: T) {.magic: "Incl".} =
     a.incl(4)
     assert a == {1, 2, 3, 4, 5}
 
+func toSet*[char|byte|bool|int16|uint16|enum](ar: openArray[T]): set[T] =
```
Comment (timotheecour) [inline lib/system/setops.nim]: we should avoid adding new symbols to `system.nim` unless properly justified.

=> move this to `std/setutils.nim` which can grow more builtin-set specific procs
(follows family of std/enumutils, etc)

Comment (timotheecour) [inline lib/system/setops.nim]: you forgot a few, and this should be exposed (also in std/setutils)
```nim
type SetElement* = char|byte|bool|int16|uint16|enum|uint8|int8`
```

File: lib/system/setops.nim [hunk 2]
```diff
@@ -9,6 +9,17 @@ func incl*[T](x: var set[T], y: T) {.magic: "Incl".} =
...
+  ##Iterates through an openArray making a set from it.
```
Comment (timotheecour) [inline lib/system/setops.nim]: just like sequtils.toSeq, you should allow the more general case of input iterator which subsubmes the `openArray` case.
this'll allow:
```nim
doAssert toSet(someIter()) == {'a', 'b'}
doAssert "ab".toSet == {'a', 'b'}
doAssert ['a','a'].toSet == {'a'}
doAssert toSet('a' .. 'c') == {'a','b','c'} # should also work with range and `HSlice`
```
etc (see how sequtils.toSeq does this)

**Comment by beef331** [comment]
Thank you for all your comments/insight, the latest commit should check all of the boxes, atleast as far as I know.

File: lib/std/setutils.nim [hunk 3]
```diff
@@ -0,0 +1,33 @@
+#
+#
+#           The Nim Compiler
+#        (c) Copyright 2020 Nim Contributors
+#
+#    See the file "copying.txt", included in this
+#    distribution, for details about the copyright.
+#
+
+## This module adds functionality to the default set
```
Comment (timotheecour) [inline lib/std/setutils.nim]: to the builtin set

File: lib/std/setutils.nim [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+
+import typetraits
+
+type SetElement* = char|byte|bool|int16|uint16|enum|uint8|int8
+
+template toSet*(iter: untyped): untyped =
+  ##Iterates through an openArray making a set from it.
+  runnableExamples:
```
Comment (timotheecour) [inline lib/std/setutils.nim:22]: this module also needs a `tests/stdlib/tsetutils.nim` where could can test more things (and use doAssert instead of assert in tsetutils)

File: lib/std/setutils.nim [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+#
+
+## This module adds functionality to the default set
+
+import typetraits
+
+type SetElement* = char|byte|bool|int16|uint16|enum|uint8|int8
+
+template toSet*(iter: untyped): untyped =
+  ##Iterates through an openArray making a set from it.
```
Comment (timotheecour) [inline lib/std/setutils.nim]: spacing + other fixes:
```
## Return a builtin set from the elements of iterable `iter`
```
(side note: later versions of nim might allow: `proc toSet*[T](iter: Iterable[T]): set[T] =`)

File: lib/std/setutils.nim [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+  runnableExamples: 
+    assert "helloWorld".toSet == {'W', 'd', 'e', 'h', 'l', 'o', 'r'}
+    assert toSet([10u16,20,30]) == {10u16, 20, 30}
+    assert [30u8,100,10].toSet == {10u8, 30, 100}
+    assert toSet(@[1321i16,321, 90]) == {90i16, 321, 1321}
+    assert toSet([false]) == {false}
+    assert toSet(0u8..10u8) == {0u8..10u8}
+  when compiles(elementType(iter)):
```
Comment (timotheecour) [inline lib/std/setutils.nim]: `compiles` has CT overhead, instead just do directly:
```nim
type E = elementType(iter)
static: doAssert E is SetElement, $E
var result: set[E]
...
```

File: lib/std/setutils.nim [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+    when elementType(iter) isnot SetElement: {.error: "Iterator does not yield a `SetElement`".}
+    else:
+      var result: set[elementType(iter)]
+      for x in iter:
+        result.incl(x)
+      result
+  else: {.error: "`toSet` can only be used on iteratable types.".}
```
Comment (timotheecour) [inline lib/std/setutils.nim]: iteratable typo but this should go away, see note above

File: lib/pure/parseutils.nim [hunk 8]
```diff
@@ -242,19 +241,9 @@ proc parseIdent*(s: string, start = 0): string =
...
-  runnableExamples:
-    var c: char
-    doAssert "nim".parseChar(c, 3) == 0
-    doAssert c == '\0'
-    doAssert "nim".parseChar(c, 0) == 1
-    doAssert c == 'n'
-  if start < s.len:
-    c = s[start]
-    result = 1
+proc parseChar*(s: string, ident: var char, start = 0): int =
```
Comment (timotheecour) [inline lib/pure/parseutils.nim]: unrelated diff; rebase from latest devel?

File: tests/stdlib/tsetutils.nim [hunk 9]
```diff
@@ -0,0 +1,5 @@
+import std/setutils
```
Comment (timotheecour) [inline tests/stdlib/tsetutils.nim:4]:
```nim
discard """
  targets: "c js"
"""

import std/setutils

template main =
  # your tests go here

main()
static: main()
```

=> so it tests in c,js,vm

File: lib/std/setutils.nim [hunk 10]
```diff
@@ -0,0 +1,34 @@
+#
+#
+#           The Nim Compiler
+#        (c) Copyright 2020 Nim Contributors
+#
+#    See the file "copying.txt", included in this
+#    distribution, for details about the copyright.
+#
+
+## This module adds functionality to the builtin set
```
Comment (timotheecour) [inline lib/std/setutils.nim]: cross reference:
```nim
## See also std/packedsets, setd/sets
```

File: lib/std/setutils.nim [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+
+import typetraits
+
+type SetElement* = char|byte|bool|int16|uint16|enum|uint8|int8
```
Comment (timotheecour) [inline lib/std/setutils.nim]: doc comment

File: lib/std/setutils.nim [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+## See also std/packedsets, setd/sets
+
+import typetraits
+
+type SetElement* = char|byte|bool|int16|uint16|enum|uint8|int8
+  ## The allowed types of a builtin set.
+
+template toSet*(iter: untyped): untyped =
+  ## Return a builtin set from the elements of iterable `iter`
+  ## Example: "ab".toSet == {'a','b'}
```
Comment (timotheecour) [inline lib/std/setutils.nim]: `Example: "ab".toSet == {'a','b'} ` is redundant, `runnableExamples:` will show usage

File: lib/std/setutils.nim [hunk 13]
```diff
@@ -0,0 +1,33 @@
...
+#
+#           The Nim Compiler
+#        (c) Copyright 2020 Nim Contributors
+#
+#    See the file "copying.txt", included in this
+#    distribution, for details about the copyright.
+#
+
+## This module adds functionality to the builtin set
+## See also std/packedsets, setd/sets
```
Comment (timotheecour) [inline lib/std/setutils.nim]: sorry, i had a typo in my review comment: 
`setd/sets` => `std/sets`

**Comment by beef331** [comment]
I'm inept with the rebase :D

**Comment by beef331** [comment]
There we go this should be fine now, thanks.

File: lib/std/setutils.nim [hunk 14]
```diff
@@ -0,0 +1,32 @@
...
+  ## Return a builtin set from the elements of iterable `iter`
+  runnableExamples: 
+    assert "helloWorld".toSet == {'W', 'd', 'e', 'h', 'l', 'o', 'r'}
+    assert toSet([10u16,20,30]) == {10u16, 20, 30}
+    assert [30u8,100,10].toSet == {10u8, 30, 100}
+    assert toSet(@[1321i16,321, 90]) == {90i16, 321, 1321}
+    assert toSet([false]) == {false}
+    assert toSet(0u8..10u8) == {0u8..10u8}
+  type E = elementType(iter)
+  static: doAssert E is SetElement, "`iter` does not yield a `SetElement`"
```
Comment (timotheecour) [inline lib/std/setutils.nim]: * you need to show `$E` somewhere in the error message. Error messages should always show context, a static string isn't enough for good debugging experience.

* but see https://github.com/nim-lang/Nim/pull/16276#discussion_r537225184

File: lib/std/setutils.nim [hunk 15]
```diff
@@ -0,0 +1,32 @@
...
+  var result: set[E]
+  for x in iter:
+    result.incl(x)
```
Comment (timotheecour) [inline lib/std/setutils.nim]: `incl(result, x)` since you're in a template, so that you get proper symbol binding.
(see manual for template + UFCS/MCS interaction).

This matters in cases where incl isn't in calling scope, eg `from system import nil` or similar

File: lib/std/setutils.nim (hunk 14 above)
Comment (timotheecour) [inline lib/std/setutils.nim]: actually, why do we even need the `static: doAssert E is SetElement` ? 
you'll still get an error when attempting `var result: set[E]` if say, E is `int64`.

File: tests/stdlib/tsetutils.nim [hunk 16]
```diff
@@ -0,0 +1,8 @@
+discard """
+  targets: "c js"
+"""
+import std/setutils
+
+doAssert "abcbb".toSet == {'a','b','c'}
+doAssert toSet([10u8,12,13]) == {10u8, 12, 13}
+doAssert toSet(0u16..30) == {0u16..30}
```
Comment (timotheecour) [inline tests/stdlib/tsetutils.nim]: add a test with `distinct`
```nim
type A = distinct char
doAssert ['x'.A].toSet == {'x'.A}
```

this should work but will fail currently in your PR because of `static: doAssert E is SetElement` 

you'd at least need `static: doAssert distinctBase(E) is SetElement`, but as I suggested above might as well simply skip this redundant check since compiler will do that for you anyways when you call `var result: set[E]`

File: lib/std/setutils.nim (hunk 14 above)
Comment (beef331) [reply lib/std/setutils.nim]: Well the main reason is to give a slightly more informative error instead of just `set is to large`

File: lib/std/setutils.nim [hunk 17]
```diff
@@ -0,0 +1,32 @@
...
+#    distribution, for details about the copyright.
+#
+
+## This module adds functionality to the builtin set
+## See also std/packedsets, set/sets
+
+import typetraits
+
+type SetElement* = char|byte|bool|int16|uint16|enum|uint8|int8
+  ## The allowed types of a builtin set.
```
Comment (timotheecour) [inline lib/std/setutils.nim]: it's a bit more complicated, because set also allows transitive closure via `distinct` and `range`.

Should we simply remove (or disable with when false to deal with this in future PR) this type, since (as I argued below) you don't even need this check:
```
static: doAssert E is SetElement, $elementType(iter) & " does not yield a `SetElement`"
```

Comment (beef331) [reply lib/std/setutils.nim]: I guess I will remove the check, should I also remove the SetElement type aswell?

File: lib/std/setutils.nim (hunk 14 above)
Comment (timotheecour) [reply lib/std/setutils.nim]: then we should fix the errmsg in the compiler, where it belongs (which would benefit other code):

```nim
nim> var z: set[int]
Error: set is too large
```

=>
```
Error: set is too large, got `int` but expected ...
```

File: lib/std/setutils.nim (hunk 17 above)
Comment (timotheecour) [reply lib/std/setutils.nim]: maybe comment it out so it's "there" for future PR's

**Review by timotheecour — APPROVED** [review]

**Comment by timotheecour** [comment]
LGTM, thanks! I restarted CI, you're hitting windows flaky test https://github.com/timotheecour/Nim/issues/411 (EDIT: reported it as https://github.com/nim-lang/Nim/issues/16317)

File: tests/stdlib/tsetutils.nim (hunk 9 above)
Comment (timotheecour) [reply tests/stdlib/tsetutils.nim:4]: @beef331 ping on this?

**Comment by timotheecour** [comment]
now you're hitting a different CI issue I haven't seen before (https://github.com/timotheecour/Nim/issues/441); I restarted failing job

**Comment by beef331** [comment]
Leave it to me to hit issues no one has hit before.

**Comment by timotheecour** [comment]
just bikeshedding but another candidate name would be:
`std/builtinsets` to avoid confusion with existing (and poorly named) `std/sets`

I'm ok with either, though

**Comment by beef331** [comment]
Do you mean `builtinsets` or `builtinsetutils`? The former may cause confusion with where `set[T]` comes from.

**Comment by timotheecour** [comment]
> The former may cause confusion with where set[T] comes from.

well by definition they're builtin ;-) ; `builtinsetutils` is a bit of a mouthful... maybe setutils is good enough

**Comment by Araq** [comment]
So ... rather than ``{'a', 'b', 'c'}`` you want to write ``toSet ['a', 'b', 'c']``? Doesn't seem useful to me. What's the use case? And now don't say "generic programming", that is **not** a use case.

**Comment by timotheecour** [comment]
> rather than {'a', 'b', 'c'} you want to write toSet ['a', 'b', 'c']? 

intended use case is not for cases where input is a litteral (in which case you'd write `{'a', 'b', 'c'}` directly), just like intended use case for `toSeq` (or `toHashSet`, etc) is not for cases where input is a literal. Instead, it's for cases where input is some iterable, which includes strings/seq/array/iterator call etc.

examples:
```nim
let input: string = getInput()
let uniqueLetters = input.toSet # or: callFn(input.toSet)
```
I've often needed this myself as well, it fits well with the rest of stdlib:
* toSeq
* toHashSet
* toPackedSet
* toIntSet
etc...

to the point that not having `toSet` seems like a weird omission

File: lib/std/setutils.nim [hunk 18]
```diff
@@ -0,0 +1,32 @@
...
+    ## The allowed types of a builtin set.
+]#
+
+template toSet*(iter: untyped): untyped =
+  ## Return a builtin set from the elements of iterable `iter`
+  runnableExamples: 
+    assert "helloWorld".toSet == {'W', 'd', 'e', 'h', 'l', 'o', 'r'}
+    assert toSet([10u16,20,30]) == {10u16, 20, 30}
+    assert [30u8,100,10].toSet == {10u8, 30, 100}
+    assert toSet(@[1321i16,321, 90]) == {90i16, 321, 1321}
```
Comment (ee7) [inline lib/std/setutils.nim]:
```suggestion
    assert toSet([10u16, 20, 30]) == {10u16, 20, 30}
    assert [30u8, 100, 10].toSet == {10u8, 30, 100}
    assert toSet(@[1321i16, 321, 90]) == {90i16, 321, 1321}
```

File: lib/std/setutils.nim [hunk 19]
```diff
@@ -0,0 +1,32 @@
...
+    assert toSet([false]) == {false}
+    assert toSet(0u8..10) == {0u8..10}
+  var result: set[elementType(iter)]
+  for x in iter:
+    incl(result,x)
```
Comment (ee7) [inline lib/std/setutils.nim]:
```suggestion
    incl(result, x)
```

File: tests/stdlib/tsetutils.nim [hunk 20]
```diff
@@ -0,0 +1,14 @@
+discard """
+  targets: "c js"
+"""
+import std/setutils
+
+template main =
+  doAssert "abcbb".toSet == {'a','b','c'}
+  doAssert toSet([10u8,12,13]) == {10u8, 12, 13}
```
Comment (ee7) [inline tests/stdlib/tsetutils.nim]:
```suggestion
  doAssert "abcbb".toSet == {'a', 'b', 'c'}
  doAssert toSet([10u8, 12, 13]) == {10u8, 12, 13}
```

File: lib/std/setutils.nim [hunk 21]
```diff
@@ -0,0 +1,32 @@
+#
+#
+#           The Nim Compiler
+#        (c) Copyright 2020 Nim Contributors
+#
+#    See the file "copying.txt", included in this
+#    distribution, for details about the copyright.
+#
+
+## This module adds functionality to the builtin set
```
Comment (ee7) [inline lib/std/setutils.nim]:
```suggestion
## This module adds functionality to the built-in `set` type.
```

File: lib/std/setutils.nim [hunk 22]
```diff
@@ -0,0 +1,32 @@
...
+## See also std/packedsets, set/sets
```
Comment (ee7) [inline lib/std/setutils.nim]:
```suggestion
## See also std/packedsets, std/sets
```
Is this a typo?

File: lib/std/setutils.nim [hunk 23]
```diff
@@ -0,0 +1,32 @@
...
+
+import typetraits
+
+#[
+  type SetElement* = char|byte|bool|int16|uint16|enum|uint8|int8
+    ## The allowed types of a builtin set.
```
Comment (ee7) [inline lib/std/setutils.nim]:
```suggestion
    ## The allowed types of a built-in set.
```

File: lib/std/setutils.nim [hunk 24]
```diff
@@ -0,0 +1,32 @@
...
+]#
+
+template toSet*(iter: untyped): untyped =
+  ## Return a builtin set from the elements of iterable `iter`
```
Comment (ee7) [inline lib/std/setutils.nim]:
```suggestion
  ## Return a built-in set from the elements of iterable `iter`
```

**Review by ee7 — COMMENTED** [review]
Some nitpicks.

**Comment by timotheecour** [comment]
@beef331 you can click "Resolve Conversation" on each item that was resolved (unless there's a disagreement) :)

**Comment by timotheecour** [comment]
hitting https://github.com/nim-lang/Nim/issues/16317 again; I've restarted failing tests

**Comment by timotheecour** [comment]
I restarted again;
you're hitting https://github.com/nim-lang/Nim/issues/16317 again (which is recently very flaky) and also https://github.com/nim-lang/Nim/issues/16315 (which has been fixed already)

This is why flaky tests really need to be fixed..

**Review by timotheecour — APPROVED** [review]

**Comment by Araq** [comment]
Needs a changelog entry and maybe (!) a ```--define:nimHasSetUtils`` (edit compiler/condsysm.nim to add it).

**Comment by timotheecour** [comment]
> Needs a changelog

indeed

> and maybe (!) a ```--define:nimHasSetUtils`` (edit compiler/condsysm.nim to add it).

this hasn't been needed for other new modules, and in fact will be obsoleted by `whichModule` https://github.com/timotheecour/Nim/issues/376 once I get to it (it's not hard); plus, `condsysm.nim` is usually about the nim binary, not the stdlib; the distinction matters because of `--lib` flag etc

File: changelog.md [hunk 25]
```diff
@@ -57,6 +57,8 @@
 
 - `strscans.scanf` now supports parsing single characters.
 
+- Added `setutils.toSet` that can take any iteratable and convert it to a built-in set,
+  if the iteratable yields a built-in settable type.
```
Comment (timotheecour) [inline changelog.md]: `iteratable` doesn't seem enligh; should be iterable

**Review by timotheecour — APPROVED** [review]

**Comment by timotheecour** [comment]
merged; previous CI was green and last commit just changed changelog.
